Mathlib/GroupTheory/GroupAction/Pointwise.lean

```diff
@@ -0,0 +1,146 @@
+
+-- @[simp] -- In #8386, the `simp_nf` linter complains:
+-- "Left-hand side does not simplify, when using the simp lemma on itself."
+-- For now we will have to manually add `image_smul_setₛₗ _` to the `simp` argument list.
+-- TODO: when lean4#3107 is fixed, mark this as `@[simp]`.
+theorem image_smul_setₛₗ [FunLike F M N] [MulActionSemiHomClass F σ M N]
+    (c : R) (s : Set M) :
+    h '' (c • s) = σ c • h '' s := by
+  simp only [← image_smul, image_image, map_smulₛₗ h]
+#align image_smul_setₛₗ image_smul_setₛₗ
+
+/-- Translation of preimage is contained in preimage of translation -/
+theorem smul_preimage_set_leₛₗ
+    [FunLike F M N] [MulActionSemiHomClass F σ M N]
+    (c : R) (t : Set N) :
+    c • h ⁻¹' t ⊆ h ⁻¹' (σ c • t) := by
+  rintro x ⟨y, hy, rfl⟩
+  exact ⟨h y, hy, by rw [map_smulₛₗ]⟩
+
+/-- General version of `preimage_smul_setₛₗ` -/
+theorem preimage_smul_setₛₗ'
+    [FunLike F M N] [MulActionSemiHomClass F σ M N]
+    (c : R) (s : Set N)
+    (hc : Function.Surjective (fun (m : M) ↦ c • m))
+    (hc' : Function.Injective (fun (n : N) ↦ σ c • n)) :
+    h ⁻¹' (σ c • s) = c • h ⁻¹' s := by
+  apply le_antisymm
+  · intro m
+    obtain ⟨m', rfl⟩ := hc m
+    rintro ⟨n, hn, hn'⟩
+    refine ⟨m', ?_, rfl⟩
+    rw [map_smulₛₗ] at hn'
+    rw [mem_preimage, ← hc' hn']
+    exact hn
+  · exact smul_preimage_set_leₛₗ M N σ h c s
+
+/-- `preimage_smul_setₛₗ` when both scalars act by unit -/
+theorem preimage_smul_setₛₗ_of_units
+    [FunLike F M N] [MulActionSemiHomClass F σ M N]
+    {c : R} (hc : IsUnit c) (hc' : IsUnit (σ c)) (s : Set N) :
+    h ⁻¹' (σ c • s) = c • h ⁻¹' s := by
+  apply preimage_smul_setₛₗ'
+  · exact (MulAction.smul_bijective_of_is_unit hc).surjective
+  · exact (MulAction.smul_bijective_of_is_unit hc').injective
+#align preimage_smul_setₛₗ preimage_smul_setₛₗ_of_units
+
+/-- `preimage_smul_setₛₗ` in the context of a `MonoidHom` -/
+theorem MonoidHom.preimage_smul_setₛₗ (σ : R →* S)
+    {F : Type*} [FunLike F M N] [MulActionSemiHomClass F ⇑σ M N] (h : F)
+    {c : R} (hc : IsUnit c) (s : Set N) :
+    h ⁻¹' (σ c • s) = c • h ⁻¹' s :=
+  preimage_smul_setₛₗ_of_units M N σ h hc (IsUnit.map σ hc) s
+
+/-- `preimage_smul_setₛₗ` in the context of a `MonoidHomClass` -/
+theorem preimage_smul_setₛₗ
+    {G : Type*} [FunLike G R S] [MonoidHomClass G R S] (σ : G)
+    [MulAction R M] [MulAction S N]
+    {F : Type*} [FunLike F M N] [MulActionSemiHomClass F σ M N] (h : F)
+    {c : R} (hc : IsUnit c) (s : Set N) :
+    h ⁻¹' (σ c • s) = c • h ⁻¹' s :=
+ MonoidHom.preimage_smul_setₛₗ _ _ (σ : R →* S) h hc s
+
+/-- `preimage_smul_setₛₗ` in the context of a groups -/
+theorem Group.preimage_smul_setₛₗ
+    {R S : Type*} [Group R] [Group S] (σ : R → S)
+    [MulAction R M] [MulAction S N]
+    {F : Type*} [FunLike F M N] [MulActionSemiHomClass F σ M N] (h : F)
+    (c : R) (s : Set N) :
+    h ⁻¹' (σ c • s) = c • h ⁻¹' s :=
+ preimage_smul_setₛₗ_of_units _ _ _ _ (Group.isUnit _) (Group.isUnit _) s
+
+variable (R)
```

Comment:
Could the remainder of this file go in a section with the appropriate `variable`s?

Reply:
Yes!